### PR TITLE
adding --sort-order option to SortSamSpark

### LINKED ISSUE #1260: SortBamSpark is missing SORT_ORDER parameter
SortBamSpark does not have a  SORT_ORDER parameter

once this is fixed - renable tests in SortBamSparkIntegrationTest

### PR DESCRIPTION
**Pull request by lbergelson**
adding a --sort-order option to SortSamSpark to let users specify the what order to sort in
enabling disabled tests
fixing the tests which weren't actually asserting anything

closes #1260

### CODE REVIEW

File: src/test/java/org/broadinstitute/hellbender/tools/spark/pipelines/SortSamSparkIntegrationTest.java [hunk 1]
```diff
@@ -38,21 +40,17 @@ public void testSortBAMs(
...
-        args.add("--output"); args.add(actualOutputFile.getCanonicalPath());
+        args.addInput(inputFile);
+        args.addOutput(actualOutputFile);
         if (null != referenceFile) {
-            args.add("--R");
-            args.add(referenceFile.getAbsolutePath());
+            args.addReference(referenceFile);
         }
-        args.add("--num-reducers"); args.add("1");
+        args.addArgument(GATKSparkTool.NUM_REDUCERS_LONG_NAME, "1");
```
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/tools/spark/pipelines/SortSamSparkIntegrationTest.java]: Why does num reducers need to be 1 for this test?

File: src/test/java/org/broadinstitute/hellbender/tools/spark/pipelines/SortSamSparkIntegrationTest.java [hunk 2]
```diff
@@ -20,9 +23,8 @@
                 {"count_reads.cram", "count_reads_sorted.cram", "count_reads.fasta", ".cram", "coordinate"},
                 {"count_reads.bam", "count_reads_sorted.bam", "count_reads.fasta", ".cram", "coordinate"},
 
-                //SortBamSpark is missing SORT_ORDER parameter  https://github.com/broadinstitute/gatk/issues/1260
-//                {"count_reads.bam", "count_reads.bam", null, ".bam", "queryname"},
-//                {"count_reads.cram", "count_reads.cram", "count_reads.fasta", ".cram", "queryname"},
+                {"count_reads.bam", "count_reads.bam", null, ".bam", "queryname"},
+                {"count_reads.cram", "count_reads.cram", "count_reads.fasta", ".cram", "queryname"},
```
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/tools/spark/pipelines/SortSamSparkIntegrationTest.java]: Is this a valid test? The inputs and outputs are the same! I'm not convinced that queryname sorting is actually working in this branch, since you use a `ReadCoordinateComparator` in the tool no matter what.

File: src/test/java/org/broadinstitute/hellbender/tools/spark/pipelines/SortSamSparkIntegrationTest.java [hunk 3]
```diff
@@ -61,13 +59,32 @@ public void test() throws Exception {
         final File sortedBam = new File(getTestDataDir(), "count_reads_sorted.bam");
         final File outputBam = createTempFile("sort_bam_spark", ".bam");
         ArgumentsBuilder args = new ArgumentsBuilder();
-        args.add("--"+ StandardArgumentDefinitions.INPUT_LONG_NAME); args.add(unsortedBam.getCanonicalPath());
-        args.add("--"+StandardArgumentDefinitions.OUTPUT_LONG_NAME); args.add(outputBam.getCanonicalPath());
-        args.add("--num-reducers"); args.add("1");
+        args.addInput(unsortedBam);
+        args.addOutput(outputBam);
+        args.addArgument(GATKSparkTool.NUM_REDUCERS_LONG_NAME, "1");
```
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/tools/spark/pipelines/SortSamSparkIntegrationTest.java]: Again, why set the number of reducers to 1?

File: src/test/java/org/broadinstitute/hellbender/tools/spark/pipelines/SortSamSparkIntegrationTest.java [hunk 4]
```diff
@@ -61,13 +59,32 @@ public void test() throws Exception {
         final File sortedBam = new File(getTestDataDir(), "count_reads_sorted.bam");
```
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/tools/spark/pipelines/SortSamSparkIntegrationTest.java]: This test method needs a better name than just `test()`. Also, why do we need this test case in addition to `testSortBAMs()`? Can they be unified?

File: src/test/resources/org/broadinstitute/hellbender/tools/count_reads_sorted.sam [hunk 5]
```diff
@@ -7,7 +7,7 @@
 @SQ	SN:chr6	LN:101
 @SQ	SN:chr7	LN:404
 @SQ	SN:chr8	LN:202
-@RG	ID:0	SM:Hi,Mom!
+@RG	ID:0	SM:Hi,Mom!	PL:ILLUMINA
```
Comment (droazen) [inline src/test/resources/org/broadinstitute/hellbender/tools/count_reads_sorted.sam]: Why did you need to make this change?

File: src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/SortSamSpark.java [hunk 6]
```diff
@@ -27,17 +29,29 @@
...
 
+    public static final String SORT_ORDER_LONG_NAME = "sort-order";
     @Override
     public boolean requiresReads() { return true; }
 
     @Argument(doc="the output file path", shortName = StandardArgumentDefinitions.OUTPUT_SHORT_NAME, fullName = StandardArgumentDefinitions.OUTPUT_LONG_NAME, optional = false)
-    protected String outputFile;
+    private String outputFile;
+
+    @Argument(doc="the order to sort the file into", fullName = SORT_ORDER_LONG_NAME, optional = true)
```
[the remark this answers is not in the document]
Comment (lbergelson) [reply src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/SortSamSpark.java]: done

File: src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/SortSamSpark.java [hunk 7]
```diff
@@ -27,17 +29,29 @@
 public final class SortSamSpark extends GATKSparkTool {
     private static final long serialVersionUID = 1L;
 
+    public static final String SORT_ORDER_LONG_NAME = "sort-order";
```
[the remark this answers is not in the document]
Comment (lbergelson) [reply src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/SortSamSpark.java]: done

File: src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/SortSamSpark.java [hunk 8]
```diff
@@ -27,17 +29,29 @@
...
+    private SAMFileHeader.SortOrder sortOrder = SAMFileHeader.SortOrder.coordinate;
 
     @Override
     public List<ReadFilter> getDefaultReadFilters() {
         return Collections.singletonList(ReadFilterLibrary.ALLOW_ALL_READS);
     }
 
+    @Override
+    protected void onStartup() {
+        if( sortOrder.getComparatorInstance() == null){
```
[the remark this answers is not in the document]
Comment (lbergelson) [reply src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/SortSamSpark.java]: fixed.  I'm sure intellij will find a different place to unbalance though

File: src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/SortSamSpark.java [hunk 9]
```diff
@@ -27,17 +29,29 @@
...
+    @Argument(doc="the order to sort the file into", fullName = SORT_ORDER_LONG_NAME, optional = true)
+    private SAMFileHeader.SortOrder sortOrder = SAMFileHeader.SortOrder.coordinate;
 
     @Override
     public List<ReadFilter> getDefaultReadFilters() {
         return Collections.singletonList(ReadFilterLibrary.ALLOW_ALL_READS);
     }
 
+    @Override
+    protected void onStartup() {
```
[the remark this answers is not in the document]
Comment (lbergelson) [reply src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/SortSamSpark.java]: On startup is documented as having an empty default implementation.  I can add it anyway though to future proof.

File: src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/SortSamSpark.java [hunk 10]
```diff
@@ -55,7 +69,7 @@ protected void runTool(final JavaSparkContext ctx) {
         } else {
             sortedReads = reads; // sorting is done by writeReads below
         }
-        readsHeader.setSortOrder(SAMFileHeader.SortOrder.coordinate);
+        readsHeader.setSortOrder(sortOrder);
```
[the remark this answers is not in the document]
Comment (lbergelson) [reply src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/SortSamSpark.java]: fixed

File: src/test/java/org/broadinstitute/hellbender/tools/spark/pipelines/SortSamSparkIntegrationTest.java (hunk 1 above)
Comment (lbergelson) [reply src/test/java/org/broadinstitute/hellbender/tools/spark/pipelines/SortSamSparkIntegrationTest.java]: I don't know...  Any number should be fine I think.

**Comment by jamesemery** [comment]
@lbergelson looks like the tests passed this time around. We should open a ticket in Hadoop-Bam to fix the issue

**Comment by lbergelson** [comment]
@jamesemery I've opened a ticket in hadoop bam.
@droazen  Could you re-review this when you get a chance?  It's super useful functionality.  If we don't trust the fix for sharded files, we could globally disable them for cram/sam instead.  I think the problem is not unique to this branch, it's just the first to test the file order or reading sharded cram/sam.

File: src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSource.java [hunk 11]
```diff
@@ -76,14 +74,22 @@ public ReadsSparkSource(final JavaSparkContext ctx, final ValidationStringency v
...
             final List<InputSplit> splits = super.getSplits(job);
-            splits.sort((a, b) -> {
-                FileSplit fa = (FileSplit) a, fb = (FileSplit) b;
-                return fa.getPath().compareTo(fb.getPath());
-            });
+
+
+            if( splits.stream().allMatch(split -> split instanceof FileVirtualSplit || split instanceof FileSplit)) {
+                splits.sort(Comparator.comparing(split -> {
+                    if (split instanceof FileVirtualSplit) {
```
Comment (jamesemery) [inline src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSource.java]: You mix the virtual and non-virtual splits here. This makes me uncomfortable as I don't know that its guaranteed either is going to return the same thing for getPath(). Could you use the filename instead? As that seems to be what is being sorted here anyway.

**Comment by jamesemery** [comment]
I'm posting an experimental version of this branch for performance purposes. This is no longer in a state where it should be merged.
